packages/provider/src/api/domainMiddleware.ts

```diff
@@ -0,0 +1,86 @@
+// Copyright 2021-2024 Prosopo (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import { validateAddress } from "@polkadot/util-crypto";
+import { ProviderApi } from "@prosopo/api";
+import { ProsopoApiError } from "@prosopo/common";
+import { ApiPrefix, DappDomainRequestBody } from "@prosopo/types";
+import type { ProviderEnvironment } from "@prosopo/types-env";
+import type { NextFunction, Request, Response } from "express";
+import { Tasks } from "../tasks/index.js";
+import { handleErrors } from "./errorHandler.js";
+
+export const domainMiddleware = (env: ProviderEnvironment) => {
+	const tasks = new Tasks(env);
+
+	return async (req: Request, res: Response, next: NextFunction) => {
+		try {
+			let parsed: { dapp: string };
+			try {
+				parsed = DappDomainRequestBody.parse(req.body);
+			} catch (err) {
+				throw siteKeyNotRegisteredError("No sitekey provided");
+			}
+			const dapp = parsed.dapp;
+
+			try {
+				validateAddress(dapp, false, 42);
+			} catch (err) {
+				throw invalidSiteKeyError(dapp);
+			}
+
+			const clientSettings = await tasks.db.getClientRecord(dapp);
+			if (!clientSettings) throw siteKeyNotRegisteredError(dapp);
+
+			const allowedDomains = clientSettings.settings?.domains;
+			if (!allowedDomains) return next(siteKeyNotRegisteredError(dapp));
+
+			const origin = req.headers.origin;
+			if (!origin) throw siteKeyNotRegisteredError(dapp);
```

Comment:
`unauthorizedOriginError` ?

Reply:
Yeah wasn't sure what error for no origin - I almost feel like telling them why it's failed does nothing but give an attacker more info

Reply:
added to unauthorizedOriginError for calls that haven't supplied origin